Unit testing and mocking (#36110)
dotnet/docs: docs/azure/sdk/unit-testing-mocking.md

---
title: Unit testing and mocking with the Azure SDK for .NET
description: Learn techniques and tools for unit testing and mocking the Azure SDK for .NET
ms.custom: devx-track-dotnet, engagement-fy23
ms.date: 07/05/2023
---

# Unit testing and mocking with the Azure SDK for .NET

> Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your apps. However, unit testing presents challenges when the code you're testing performs network calls to Azure resources. Tests that run against live services can experience issues, such as latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.

Reviewer: ```suggestion
Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your apps. However, unit testing presents challenges when the code you're testing performs network calls, such as those made to Azure resources. Tests that run against live services can experience issues, such as latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.
```

In this article, you'll learn how to write unit tests for the Azure SDK for .NET that isolate your dependencies to make your tests more reliable. You'll also learn how to replace key components with in-memory test implementations to create fast and reliable unit tests, and see how to design your own classes to better support unit testing. This article includes examples that use [Moq](https://www.nuget.org/packages/moq/), which is a popular mocking and testing library for .NET.

## Understand service clients

A service client class is the main entry point for developers in an Azure SDK library and implements most of the logic to communicate with the Azure service. When unit testing service client classes, it's important to be able to create an instance of the client that behaves as expected without making any network calls.

Each of the Azure SDK clients follows [mocking guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-mocking) that allow their behavior to be overridden:

* Each client offers at least one protected constructor to allow inheritance for testing.
* All public client members are virtual to allow overriding.

> [!NOTE]

> > The code examples in this article use types from the [`Azure.Security.KeyVault.Secrets`](https://www.nuget.org/packages/Azure.Security.KeyVault.Secrets/) library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.

Reviewer: ```suggestion
> The code examples in this article use types from the [Azure.Security.KeyVault.Secrets](https://www.nuget.org/packages/Azure.Security.KeyVault.Secrets/) library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.
```

To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).

> # [Non-library](#tab/csharp)

Reviewer: I'd update all of these tabs to be `non-library`, instead of `csharp`.

---
title: Unit testing and mocking with the Azure SDK for .NET
description: Learn techniques and tools for unit testing and mocking the Azure SDK for .NET
ms.custom: devx-track-dotnet, engagement-fy23
ms.date: 07/05/2023
---

# Unit testing and mocking with the Azure SDK for .NET

Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your apps. However, unit testing presents challenges when the code you're testing performs network calls, such as those made to Azure resources. Tests that run against live services can experience issues, such as latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.

> In this article, you'll learn how to write unit tests for the Azure SDK for .NET that isolate your dependencies to make your tests more reliable. You'll also learn how to replace key components with in-memory test implementations to create fast and reliable unit tests, and see how to design your own classes to better support unit testing. This article includes examples that use [Moq](https://www.nuget.org/packages/moq/) and [NSubstitute](https://www.nuget.org/packages/nsubstitute/), which are popular mocking and testing libraries for .NET.

Reviewer 2: ```suggestion
In this article, you'll learn how to write unit tests for the Azure SDK for .NET that isolate your dependencies to make your tests more reliable. You'll also learn how to replace key components with in-memory test implementations to create fast and reliable unit tests, and see how to design your own classes to better support unit testing. This article includes examples that use [Moq](https://www.nuget.org/packages/moq/) and [NSubstitute](https://www.nuget.org/packages/nsubstitute/), which are popular mocking libraries for .NET.
```

:::code language="csharp" source="snippets/unit-testing/NonLibrary/MockSecretClient.cs" :::

# [Moq](#tab/moq)

:::code language="csharp" source="snippets/unit-testing/Moq/TestSnippets_Moq.cs" id="MockSecretClient" :::

# [NSubstitute](#tab/nsubstitute)

:::code language="csharp" source="snippets/unit-testing/NSubstitute/TestSnippets_NSubstitute.cs" id="MockSecretClient" :::

---

### Service client input and output models

Model types hold the data being sent and received from Azure services. There are three types of models:

* **Input models** are intended to be created and passed as parameters to service methods by developers. They have one or more public constructors and writeable properties.
* **Output models** are only returned by the service and have neither public constructors nor writeable properties.
* **Round-trip models** are less common, but are returned by the service, modified, and used as an input.

To create a test instance of an input model use one of the available public constructors and set the additional properties you need.

```csharp
var secretProperties = new SecretProperties("secret")
{
    NotBefore = DateTimeOffset.Now
};
```

> To create instances of output models, a model factory is used. Most Azure SDK client libraries provide a static model factory class that ends in `ModelFactory` and contains a set of static methods to initialize the library's output model types.

Reviewer 2: Feedback addressed over in https://github.com/dotnet/docs/pull/36339

---
title: Unit testing and mocking with the Azure SDK for .NET
description: Learn techniques and tools for unit testing and mocking the Azure SDK for .NET
ms.custom: devx-track-dotnet, engagement-fy23
ms.date: 07/05/2023
---

# Unit testing and mocking with the Azure SDK for .NET

Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your apps. However, unit testing presents challenges when the code you're testing performs network calls, such as those made to Azure resources. Tests that run against live services can experience issues, such as latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.

> In this article, you'll learn how to write unit tests for the Azure SDK for .NET that isolate your dependencies to make your tests more reliable. You'll also learn how to replace key components with in-memory test implementations to create fast and reliable unit tests, and see how to design your own classes to better support unit testing. This article includes examples that use [Moq](https://www.nuget.org/packages/moq/), which is a popular mocking and testing library for .NET.

Reviewer 2: Now that we have NSubstitute code samples, mention that library in the last sentence too.

> [!NOTE]
> It can be cumbersome to manually define test classes, especially if you need to customize behavior differently for each test. Consider using a library like Moq to streamline your testing.

:::code language="csharp" source="snippets/unit-testing/NonLibrary/MockSecretClient.cs" :::

# [Moq](#tab/moq)

:::code language="csharp" source="snippets/unit-testing/Moq/TestSnippets_Moq.cs" id="MockSecretClient" :::

# [NSubstitute](#tab/nsubstitute)

:::code language="csharp" source="snippets/unit-testing/NSubstitute/TestSnippets_NSubstitute.cs" id="MockSecretClient" :::

---

### Service client input and output models

Model types hold the data being sent and received from Azure services. There are three types of models:

* **Input models** are intended to be created and passed as parameters to service methods by developers. They have one or more public constructors and writeable properties.
* **Output models** are only returned by the service and have neither public constructors nor writeable properties.

> * **Round-trip** models are less common, but are returned by the service, modified, and used as an input.

Reviewer 2: ```suggestion
* **Round-trip models** are less common, but are returned by the service, modified, and used as an input.
```

To create a test instance of an input model use one of the available public constructors and set the additional properties you need.

```csharp

> SecretProperties secretProperties = new("secret")

Reviewer 2: ```suggestion
var secretProperties = new SecretProperties("secret")
```

{
    NotBefore = DateTimeOffset.Now
};
```

To create instances of output models, a model factory is used. Most Azure SDK client libraries provide a static model factory class that ends in `ModelFactory` and contains a set of static methods to initialize the library's output model types.

```csharp
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
```

> [!NOTE]
> Some input models have read-only properties that are only populated when the model is returned by the service. In this case, a model factory method will be available that allows setting these properties. For example, <xref:Azure.Security.KeyVault.Secrets.SecretModelFactory.SecretProperties%2A>.

```csharp
// CreatedOn is a read-only property and can only be 
// set via the model factory's SecretProperties method.
secretPropertiesWithCreatedOn = SecretModelFactory.SecretProperties(
    name: "secret", createdOn: DateTimeOffset.Now);
```

## Explore response types

> The <xref:Azure.Response> class is an abstract class that represents an HTTP response and is returned by most service client methods. You can create test `Response` instances using either the Moq library or standard C# inheritance.

Reviewer 2: Mention NSubstitute here too

---

Some services also support using the <xref:Azure.Response%601> type, which is a class that contains a model and the HTTP response that returned it. To create a test instance of `Response<T>`, use the static `Response.FromValue` method:

## [Non-library](#tab/csharp)

:::code language="csharp" source="snippets/unit-testing/NonLibrary/TestSnippets.cs" id="MockResponseT" :::

## [Moq](#tab/moq)

:::code language="csharp" source="snippets/unit-testing/Moq/TestSnippets_Moq.cs" id="MockResponseT" :::

# [NSubstitute](#tab/nsubstitute)

:::code language="csharp" source="snippets/unit-testing/NSubstitute/TestSnippets_NSubstitute.cs" id="MockResponseT" :::

---

### Explore paging

The <xref:Azure.Page%601> class is used as a building block in service methods that invoke operations returning results in multiple pages. The `Page<T>` is rarely returned from APIs directly but is useful to create the `AsyncPageable<T>` and `Pageable<T>` instances in the next section. To create a `Page<T>` instance, use the `Page<T>.FromValues` method, passing a list of items, a continuation token, and the `Response`.

> The `continuationToken` parameter is used to retrieve the next page from the service. For unit testing purposes, it should be set to null for the last page and should be non-empty for other pages.

Reviewer 2: ```suggestion
The `continuationToken` parameter is used to retrieve the next page from the service. For unit testing purposes, it should be set to `null` for the last page and should be non-empty for other pages.
```

To create a test instance of `Pageable` or `AsyncPageable`, use the `FromPages` static method:

## [Non-library](#tab/csharp)

:::code language="csharp" source="snippets/unit-testing/NonLibrary/TestSnippets.cs" id="MultipleResponsePage" :::

## [Moq](#tab/moq)

:::code language="csharp" source="snippets/unit-testing/Moq/TestSnippets_Moq.cs" id="MultipleResponsePage" :::

# [NSubstitute](#tab/nsubstitute)

:::code language="csharp" source="snippets/unit-testing/NSubstitute/TestSnippets_NSubstitute.cs" id="MultipleResponsePage" :::

---

## Write a mocked unit test

Suppose your app contains a class that finds the names of keys that will expire within a given amount of time.

:::code language="csharp" source="snippets/unit-testing/AboutToExpireSecretsFinder.cs" :::

You want to test the following behaviors of the `AboutToExpireSecretFinder` to ensure they continue working as expected:

* Secrets without an expiry date set aren't returned.
* Secrets with an expiry date closer to the current date than the threshold are returned.

> When unit testing you only want the unit tests to verify the application logic and not whether the Azure service or library works correctly. The following example tests the key behaviors using the popular xUnit framework for C#:

Reviewer 2: ```suggestion
When unit testing you only want the unit tests to verify the application logic and not whether the Azure service or library works correctly. The following example tests the key behaviors using the popular [xUnit](https://www.nuget.org/packages/xunit) library:
```

---
title: Unit testing and mocking with the Azure SDK .NET
description: Learn techniques and tools for unit testing and mocking the Azure SDK for .NET
ms.custom: devx-track-dotnet, engagement-fy23
ms.date: 07/05/2023
---

# Unit testing and mocking with the Azure SDK .NET

> Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your app. However, unit testing presents challenges when the code you are testing communicates with an Azure service over a network. Tests that run against live services can experience issues such as latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.

Reviewer: ```suggestion
Unit testing is an important part of a sustainable development process that can improve code quality and prevent regressions or bugs in your apps. However, unit testing presents challenges when the code you're testing performs network calls to Azure resources. Tests that run against live services can experience issues, such as, latency that slows down test execution, dependencies on code outside of the isolated test, and issues with managing service state and costs every time the test is run. Instead of testing against live Azure services, replace the service clients with mocked or in-memory implementations. This avoids the above issues and lets developers focus on testing their application logic, independent from the network and service.
```

> In this article, we’ll show you how to use the Azure SDK to write great unit tests that isolate your dependencies to make your tests more reliable. First, we’ll show you how many of the Azure SDK building blocks have been designed to be replaced with an in-memory test implementation. Then, we’ll show you how the pieces come together into a fast and reliable unit test. Finally, we’ll provide some design tips to help you design your own classes to better support unit testing.

Reviewer: Avoid "we", you're not doing this with the reader... 🙂 Also, seeing smart quotes, instead use normal quotes (was this copied from somewhere). You can turn on "Replace Smart Quotes" option that I wrote for the Docs Authoring Pack extension:

![image](https://github.com/dotnet/docs/assets/7679720/70649e5f-ff87-409d-a919-03f32ed81ba1)

```suggestion
In this article, you'll learn how to use the Azure SDK to write great unit tests that isolate your dependencies to make your tests more reliable. First, you'll see how many of the Azure SDK building blocks have been designed to be replaced with an in-memory test implementation. Then, you'll see you how the pieces come together into a fast and reliable unit test. Finally, you'll learn some design tips to help you design your own classes to better support unit testing.
```

> ## Understanding service clients

Reviewer: Per style guide, avoid gerunds in headings.

```suggestion
## Understand service clients
```

A service client class is the main entry point for developers in an Azure SDK library and implements most of the logic to communicate with the Azure service. When unit testing service client classes, it’s important to be able to create an instance of the client that behaves as expected without making any network calls.

Each of the Azure SDK clients follows mocking guidelines that allow their behavior to be overridden:

* Each client offers at least one protected constructor to allow inheritance for testing.
* All public client members are virtual to allow overriding.

> [!NOTE]
> The code examples in this article use types from the `Azure.Security.KeyVault.Secrets` library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.

To create a test client instance, inherit from the client type and override methods you are calling in your code with an implementation that returns a set of test objects. Most clients contain both synchronous and asynchronous methods for operations; override only the one your application code is calling.

```csharp

> public class MockSecretClient : SecretClient 
> { 
>     public MockSecretClient() {}
> 
>     public override Response<KeyVaultSecret> GetSecret(string name, string version = null, CancellationToken cancellationToken = default) => ...;
>     public override Task<Response<KeyVaultSecret>> GetSecretAsync(string name, string version = null, CancellationToken cancellationToken = default) => ...;
> 

Reviewer: An empty `.ctor` is redundant, also, lots of folks will Copy this code - and if it doesn't compile that's problematic. Instead of `...` (invalid code), `=> throw new NotImplementedException()`. Also, please try to avoid horizontal scrolling by formatting code for vertical scrolling.

```suggestion
public class MockSecretClient : SecretClient 
{
    public override Response<KeyVaultSecret> GetSecret(
        string name,
        string version = null, 
        CancellationToken cancellationToken = default)
        => throw new NotImplementedException();
        
    public override Task<Response<KeyVaultSecret>> GetSecretAsync(
        string name, 
        string version = null, 
        CancellationToken cancellationToken = default)
        => throw new NotImplementedException();
```

}

SecretClient secretClient = new MockSecretClient();
```

It can be cumbersome to define a test instance of the class, especially if you need to customize behavior differently for each test. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article). We’ll illustrate this set of examples using a popular .NET mocking framework, Moq.

To create a test client instance using Moq:

```csharp

> Mock<SecretClient> clientMock = new Mock<SecretClient>();
> clientMock.Setup(c => c.GetSecret(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<CancellationToken>()))
>         .Returns(...);
> 
> clientMock.Setup(c => c.GetSecretAsync(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<CancellationToken>()))
>         .ReturnsAsync(...);
> 
> SecretClient secretClient = clientMock.Object;

Reviewer: ```suggestion
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
Mock<SecretClient> clientMock = new Mock<SecretClient>();
clientMock.Setup(c => c.GetSecret(
        It.IsAny<string>(),
        It.IsAny<string>(),
        It.IsAny<CancellationToken>())
    )
    .Returns(Response.FromValue(keyVaultSecret, Mock.Of<Response>()));

clientMock.Setup(c => c.GetSecretAsync(
        It.IsAny<string>(), 
        It.IsAny<string>(), 
        It.IsAny<CancellationToken>())
    )
    .ReturnsAsync(Response.FromValue(keyVaultSecret, Mock.Of<Response>()));

SecretClient secretClient = clientMock.Object;
```

```

> [!NOTE]
> When created using a parameterless constructor, the client is not fully initialized leaving client behavior undefined. In practice, this means that most will throw an exception if called without an overridden implementation.

## Working with input and output models

Model types hold the data being sent and received from Azure services. There are two main kinds of models.

* **Input models** are intended to be created and passed as parameters to service methods by developers. They have one or more public constructors and writeable properties.
* **Output models** are only returned by the service and have neither public constructors nor writeable properties.

To create a test instance of an input model use one of the available public constructors and set additional properties you need.

```csharp

> SecretProperties secretProperties = new SecretProperties("secret"); secretProperties.NotBefore = DateTimeOffset.Now;

Reviewer: ```suggestion
SecretProperties secretProperties = new("secret")
{
    NotBefore = DateTimeOffset.Now
};
```

```

To create instances of output models, a model factory is used. For most Azure SDK client libraries, the model factory is a static class that ends in ModelFactory and contains a set of static methods to create and initialize the library’s output model types.

```C#

> KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(new SecretProperties("secret"), "secretValue");

Reviewer: ```suggestion
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
```

```

> [!NOTE]

> > Some input models have read-only properties that are only populated when the model is returned by the service. In this case, a model factory method will be available that allows setting these properties.

Reviewer: ```suggestion
> Some input models have read-only properties that are only populated when the model is returned by the service. In this case, a model factory method will be available that allows setting these properties. See <xref:Azure.Security.KeyVault.Secrets.SecretModelFactory>.
```

```csharp

> // CreatedOn is a read-only property and can only be set via a model factory SecretProperties        
> secretPropertiesWithCreatedOn = SecretModelFactory.SecretProperties(name: "secret", createdOn: DateTimeOffset.Now);

Reviewer: ```suggestion
// CreatedOn is a read-only property and can only be 
// set via a model factory SecretProperties.
secretPropertiesWithCreatedOn = SecretModelFactory.SecretProperties(
    name: "secret", createdOn: DateTimeOffset.Now);
```

```

> ## Working with response types

Reviewer: ```suggestion
## Work with response types
```

> The `Response` class is an abstract class that represents an HTTP response and is a part of almost all types returned by client methods. There are several different variations of the `Response` type you will encounter when testing Azure services. You can create test `Response` instances using either the Moq library or standard C# inheritence.

Reviewer: First reference to `Response` should be an xref.

## [Moq](#tab/moq)

The Moq library provides concise functionality for setting up mock responses:

```csharp
Mock responseMock = new Mock();
responseMock.SetupGet(r => r.Status).Returns(200);

Response response = responseMock.Object;
```

To create an instance of Response without defining any behaviors:

```csharp
Response response = Mock.Of<Response>();
```

## [C#](#tab/csharp)

> The Response class is abstract, which means there are a lot of members to override. Consider using the Moq library to streamline your approach.

Reviewer: Please link to the Moq NuGet package.

```csharp
public class TestResponse : Response
{

>     protected override bool TryGetHeader(string name, out string value) => ...;
>     protected override bool TryGetHeaderValues(string name, out IEnumerable<string> values) => ...;
>     protected override bool ContainsHeader(string name) => ...;
>     protected override IEnumerable<HttpHeader> EnumerateHeaders() => ...;
> 
>     public override int Status => ...;
>     public override string ReasonPhrase => ...;
>     public override Stream ContentStream => ...;
>     public override string ClientRequestId => ...;
> 
>     public override void Dispose() => ...;

Reviewer: ```suggestion
    protected override bool TryGetHeader(string name, out string value)
        => throw new NotImplementedException();
    protected override bool TryGetHeaderValues(string name, out IEnumerable<string> values)
        => throw new NotImplementedException();
         
    protected override bool ContainsHeader(string name) => throw new NotImplementedException();
    protected override IEnumerable<HttpHeader> EnumerateHeaders() 
        => throw new NotImplementedException();

    public override int Status => throw new NotImplementedException();
    public override string ReasonPhrase => throw new NotImplementedException();
    public override Stream ContentStream => throw new NotImplementedException();
    public override string ClientRequestId => throw new NotImplementedException();

    public override void Dispose() => throw new NotImplementedException();
```

}
```

---

> ### Response<T>

Reviewer: ```suggestion
### Response\<T\>
```

Or

```md
### `Response<T>`
```

> The `Response<T>` is a class that contains a model and the HTTP response that returned it. To create a test instance of `Response<T>` use the static Response.FromValue method:

Reviewer: xref

## [Moq](#tab/moq)

```csharp

> KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(new SecretProperties("secret"), "secretValue");
> Response response = Response.FromValue(keyVaultSecret, new TestResponse());

Reviewer: ```suggestion
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
Response response = Response.FromValue(keyVaultSecret, new TestResponse());
```

```

## [C#](#tab/csharp)

```csharp

> KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(new SecretProperties("secret"), "secretValue");
> Response<KeyVaultSecret> response = Response.FromValue(keyVaultSecret, Mock.Of<Response>());

Reviewer: ```suggestion
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
Response<KeyVaultSecret> response = Response.FromValue(
    keyVaultSecret, Mock.Of<Response>());
```

```

---

> ### Page<T>

Reviewer: ```suggestion
### Page\<T\>
```

Or 

```md
### `Page<T>`
```

> The Page<T> is used as a building block in service methods that invoke operations returning results in multiple pages. The Page<T> is rarely returned from APIs directly but is useful to create the AsyncPageable<T> and Pageable<T> instances we’ll discuss in the next section. To create a Page<T> instance, use the Page<T>.FromValues method, passing a list of items, a continuation token, and the Response.

Reviewer: Alos, please add xref.

```suggestion
The `Page<T>` is used as a building block in service methods that invoke operations returning results in multiple pages. The `Page<T>` is rarely returned from APIs directly but is useful to create the `AsyncPageable<T>` and `Pageable<T>` instances we’ll discuss in the next section. To create a `Page<T>` instance, use the `Page<T>.FromValues` method, passing a list of items, a continuation token, and the `Response`.
```

> The continuationToken parameter is used to retrieve the next page from the service. For unit testing purposes, it should be set to null for the last page and should be non-empty for other pages.

Reviewer: ```suggestion
The `continuationToken` parameter is used to retrieve the next page from the service. For unit testing purposes, it should be set to null for the last page and should be non-empty for other pages.
```

## [Moq](#tab/moq)

```csharp
Page<SecretProperties> responsePage = Page<SecretProperties>.FromValues(
    new[] {
        new SecretProperties("secret1"),
        new SecretProperties("secret2")
    },
    continuationToken: null,
    Mock.Of<Response>());
```

## [C#](#tab/csharp)

```csharp

> Page responsePage = Page.FromValues( new[] { new SecretProperties("secret1"), new SecretProperties("secret2") }, continuationToken: null, new TestResponse());

Reviewer: ```suggestion
Page responsePage = Page.FromValues(
    new[]
    {
        new SecretProperties("secret1"),
        new SecretProperties("secret2")
    }, 
    continuationToken: null,
    new TestResponse());
```

```

---

> ### AsyncPageable<T> and Pageable<T>

Reviewer: ```suggestion
### AsyncPageable\<T\> and Pageable\<T\>
```

Or

```md
### `AsyncPageable<T>` and `Pageable<T>`
```

> AsyncPageable<T> and Pageable<T> are classes that represent collections of models returned by the service in pages. The only difference between them is that one is used with synchronous methods while the other is used with asynchronous methods.

Reviewer: ```suggestion
`AsyncPageable<T>` and `Pageable<T>` are classes that represent collections of models returned by the service in pages. The only difference between them is that one is used with synchronous methods while the other is used with asynchronous methods.
```

To create a test instance of Pageable or AsyncPageable, use the FromPages static method:

```csharp

> Page page1 = Page.FromValues(new[] { new SecretProperties("secret1"), new SecretProperties("secret2") }, "continuationToken", Mock.Of<Response>());
> 
> Page page2 = Page.FromValues(new[] { new SecretProperties("secret3"), new SecretProperties("secret4") }, "continuationToken2", Mock.Of<Response>());
> 
> Page lastPage = Page.FromValues(new[] { new SecretProperties("secret5"), new SecretProperties("secret6") }, continuationToken: null, Mock.Of<Response>());
> 
> Pageable pageable = Pageable.FromPages(new[] { page1, page2, lastPage });
> 
> AsyncPageable asyncPageable = AsyncPageable.FromPages(new[] { page1, page2, lastPage }); )

Reviewer: ```suggestion
Page page1 = Page.FromValues(
    new[]
    {
        new SecretProperties("secret1"),
        new SecretProperties("secret2")
    },
    "continuationToken",
    Mock.Of<Response>());

Page page2 = Page.FromValues(
    new[]
    {
        new SecretProperties("secret3"),
        new SecretProperties("secret4")
    },
    "continuationToken2",
    Mock.Of<Response>());

Page lastPage = Page.FromValues(
    new[]
    {
        new SecretProperties("secret5"),
        new SecretProperties("secret6")
    },
    continuationToken: null,
    Mock.Of<Response>());

Pageable pageable = Pageable.FromPages(new[] { page1, page2, lastPage });

AsyncPageable asyncPageable = AsyncPageable.FromPages(new[] { page1, page2, lastPage }); )
```

```

> ## Writing a mocked unit test

Reviewer: ```suggestion
## Write a mocked unit test
```

Suppose your class contains a class that finds the names of keys that will expire within a given amount of time.

```csharp
public class AboutToExpireSecretFinder
{
    private readonly TimeSpan _threshold;
    private readonly SecretClient _client;

    public AboutToExpireSecretFinder(TimeSpan threshold, SecretClient client)
    {
        _threshold = threshold;
        _client = client;
    }

    public async Task<string[]> GetAboutToExpireSecrets()
    {

>         List<string> secretsAboutToExpire = new List<string>();

Reviewer: ```suggestion
        List<string> secretsAboutToExpire = new();
```

        await foreach (var secret in _client.GetPropertiesOfSecretsAsync())
        {

>             if (secret.ExpiresOn != null &&
>                 secret.ExpiresOn.Value - DateTimeOffset.Now <= _threshold)

Reviewer: ```suggestion
            if (secret.ExpiresOn.HasValue &&
                secret.ExpiresOn.Value - DateTimeOffset.Now <= _threshold)
```

            {
                secretsAboutToExpire.Add(secret.Name);
            }
        }

        return secretsAboutToExpire.ToArray();
    }
}
```

You want to test the following behaviors of the `AboutToExpireSecretFinder` to ensure they continue working as expected:

* Secrets that don’t have an expiry date set are not returned.
* Secrets with an expiry date closer to the current date than the threshold are returned.

When unit testing you only want the unit tests to verify the application logic and not whether the Azure service or SDK works correctly. The following example tests the key behaviors using the popular xUnit framework for C#:

```csharp

> public class AboutToExpireSecretFinderTests { 

Reviewer: ```suggestion
public class AboutToExpireSecretFinderTests
{
```

    [Fact] 
    public async Task DoesNotReturnNonExpiringSecrets() 

>     { // Arrange
>     

Reviewer: ```suggestion
    {
        // Arrange
```

        {
            new SecretProperties("secret1") { ExpiresOn = now.AddDays(1) },
            new SecretProperties("secret2") { ExpiresOn = now.AddDays(2) },
            new SecretProperties("secret3") { ExpiresOn = now.AddDays(3) }
        }, null, Mock.Of<Response>());
    
        // Create a pageable that consists of a single page
        AsyncPageable<SecretProperties> pageable = AsyncPageable<SecretProperties>.FromPages(new [] { page });
    
        // Setup a client mock object to return the pageable when GetPropertiesOfSecretsAsync is called
        var clientMock = new Mock<SecretClient>();
        clientMock.Setup(c => c.GetPropertiesOfSecretsAsync(It.IsAny<CancellationToken>()))
            .Returns(pageable);
    
        // Create an instance of a class to test passing in the mock client
        var finder = new AboutToExpireSecretFinder(TimeSpan.FromDays(2), clientMock.Object);
    
        // Act
        var soonToExpire = await finder.GetAboutToExpireSecrets();
    
        // Assert
        Assert.Equal(new[] {"secret1", "secret2"}, soonToExpire);
    }
}
```

## Refactoring your types for testability

> Classes that need to be tested should be designed with the ability to replace their dependency implementations. It was a seamless process to replace the `SecretClient` implementation in the example from the previous section because it was one of the constructor parameters. However, there may be classes in your code that create their own dependencies and are not easily testable, such as the following:

Reviewer: xref please

```csharp
public class AboutToExpireSecretFinder
{
    public AboutToExpireSecretFinder(TimeSpan threshold)
    {
        _threshold = threshold;

>         _client = new SecretClient(new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")), new DefaultAzureCredential());

Reviewer: ```suggestion
        _client = new SecretClient(
            new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")),
            new DefaultAzureCredential());
```

    }
}
```

> The simplest refactoring you can do to enable testing with dependency injection would be to expose the client as a parameter and run default creation code when no value is provided. This approach allows you to make the class testable while still retaining the flexibility of using the type without much ceremony.

Reviewer: Might make sense to add a cross reference article link to the DI with Azure content.

```csharp
public class AboutToExpireSecretFinder
{
    public AboutToExpireSecretFinder(TimeSpan threshold, SecretClient client = null) 
    { 

>         \_threshold = threshold; 
>         \_client = client ?? new SecretClient(new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")), new DefaultAzureCredential());

Reviewer: ```suggestion
        _threshold = threshold; 
        _client = client ?? new SecretClient(
            new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")),
            new DefaultAzureCredential());
```

    }
}
```

Another option is to move the dependency creation entirely into the calling code:

```csharp
public class AboutToExpireSecretFinder
{
    public AboutToExpireSecretFinder(TimeSpan threshold, SecretClient client)
    {
        _threshold = threshold;
        _client = client;
    }
}

> var secretClient = new SecretClient(new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")), new DefaultAzureCredential());
> var finder = new AboutToExpireSecretFinder(TimeSpan.FromDays(2), secretClient);

Reviewer: ```suggestion
var secretClient = new SecretClient(
    new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")), 
    new DefaultAzureCredential());
var finder = new AboutToExpireSecretFinder(TimeSpan.FromDays(2), secretClient);
```

```

This approach is useful when you would like to consolidate the dependency creation and share the client between multiple consuming classes.

> > [!NOTE]

Reviewer: Might be better to include a See also section with related content.

---

> title: Unit testing and mocking with the Azure SDK for .NET

Reviewer 3: Should we reference this older post on Mocking https://devblogs.microsoft.com/azure-sdk/unit-testing-and-mocking/ and only focus on the new and/or improved bits?

Reviewer 2: My vote is no. Once this doc publishes, I'd like to do these 2 things:

1. Add a note at the top of that blog post that directs customers to the doc for the latest guidance.
2. Replace the content at https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/Mocking.md with a link to the doc. The doc will be much more discoverable and localized.

In this article, you'll learn how to write unit tests for the Azure SDK for .NET that isolate your dependencies to make your tests more reliable. You'll also learn how to replace key components with in-memory test implementations to create fast and reliable unit tests, and see how to design your own classes to better support unit testing. This article includes examples that use [Moq](https://www.nuget.org/packages/moq/), which is a popular mocking and testing library for .NET.

## Understand service clients

A service client class is the main entry point for developers in an Azure SDK library and implements most of the logic to communicate with the Azure service. When unit testing service client classes, it's important to be able to create an instance of the client that behaves as expected without making any network calls.

> Each of the Azure SDK clients follows mocking guidelines that allow their behavior to be overridden:

Reviewer 4: ```suggestion
Each of the Azure SDK clients follows [mocking guidelines](https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-mocking) that allow their behavior to be overridden:
```

* Each client offers at least one protected constructor to allow inheritance for testing.
* All public client members are virtual to allow overriding.

> [!NOTE]
> The code examples in this article use types from the [`Azure.Security.KeyVault.Secrets`](https://www.nuget.org/packages/Azure.Security.KeyVault.Secrets/) library for the Azure Key Vault service. The concepts demonstrated in this article also apply to service clients from many other Azure services, such as Azure Storage or Azure Service Bus.

> To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).

Reviewer 5: 💡 I would consider both options here code smells. Where possible tests should be operating under production code to the maximum degree possible. When this involves hosted APIs that are not available (or not practical) during test scenarios, the preferred approach is to spin up a stub implementation of the actual service using something like self-hosted OWIN, and direct the production code to use that local endpoint as the target for its operations.

Author: In the complete unit test example at the end of the article the only part of the code that is being mocked or substituted are the SDK classes - the production code written for the app by the developer is still being tested. The SDK dependencies are being swapped with mocks because a unit test wouldn't be concerned with testing the code in external library dependencies - this was the guidance provided by the Azure SDK team when this content was requested.

Reviewer 5: ➡️ Will discuss with content authors internally to find a documentation plan that naturally encourages the use of tools for specific areas where they excel, and can't be interpreted as suggesting and single tool as a one-size-fixes-all solution.

Reviewer: Hi @sharwell - I'd like to better understand your arguments.

This article isn't trying to sell Moq as a "one-size-fits-all solution", it's celebrating it as an OSS project that works well with the Azure .NET SDK. It's got nearly 500m downloads on NuGet and is used by 175k other public GitHub repositories for a reason.

Moq is used in the dotnet/runtime too, and it's perfectly valid when you know how to use the tool. The purpose of this article is to demonstrate how to use it.

What specifically about this article is off putting, or what specifically would you suggest changing? It's based on this blog post: https://devblogs.microsoft.com/azure-sdk/unit-testing-and-mocking.

Reviewer 6: @sharwell this article follows our Azure SDK for .NET guidelines: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-mocking /cc @KrzysztofCwalina

> To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).

Reviewer 5: 💡 Mocking libraries like Moq are especially harmful to the health of a test suite when used at scale. Since they do not follow the inheritance rules defined by the language, features such as Go To Implementation will not correctly identify them as implementations of a member when developers are working to update their code. This can lead to large numbers of false positive test failures in non-breaking development scenarios.

Reviewer 5: ➡️ Absorbed by https://github.com/dotnet/docs/pull/36110#discussion_r1255846092

> To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).

Reviewer 6: Tip: second time you referred to this. You can use page links (typically at the bottom, but don't have to be) and if the anchor text matches example, you don't even need to use a separate set of brackets:

````markdown
This [Moq][Moq] link renders the same as this [Moq] link.

[Moq]: https://nuget.org/packages/moq

> To create a test service client, you can either use a mocking library such as [Moq](https://www.nuget.org/packages/moq/) or standard C# features such as inheritance. Mocking frameworks allow you to simplify the code that you must write to override member behavior (as well as other useful features that are beyond the scope of this article).

Reviewer 2: After today's meeting, we agreed to also add tabs for NSubstitute code samples. Again, the thought here is we don't want to over-endorse one particular approach.

# [Moq](#tab/moq)

```csharp
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
Mock<SecretClient> clientMock = new Mock<SecretClient>();
clientMock.Setup(c => c.GetSecret(
        It.IsAny<string>(),
        It.IsAny<string>(),
        It.IsAny<CancellationToken>())
    )
    .Returns(Response.FromValue(keyVaultSecret, Mock.Of<Response>()));

clientMock.Setup(c => c.GetSecretAsync(
        It.IsAny<string>(), 
        It.IsAny<string>(), 
        It.IsAny<CancellationToken>())
    )
    .ReturnsAsync(Response.FromValue(keyVaultSecret, Mock.Of<Response>()));

SecretClient secretClient = clientMock.Object;
```

> # [C#](#tab/csharp)

Reviewer 2: Let's swap the order of the 2 tabs. We don't want it to seem as though we're overly endorsing Moq.

> To create a test client instance using C# without a library, inherit from the client type and override methods you are calling in your code with an implementation that returns a set of test objects. Most clients contain both synchronous and asynchronous methods for operations; override only the one your application code is calling.

Reviewer 7: ```suggestion
To create a test client instance using C# without a mocking library, inherit from the client type and override methods you are calling in your code with an implementation that returns a set of test objects. Most clients contain both synchronous and asynchronous methods for operations; override only the one your application code is calling.
```

> [!NOTE]
> It can be cumbersome to manually define test classes, especially if you need to customize behavior differently for each test. Consider using a library like Moq to streamline your testing.

```csharp
public class MockSecretClient : SecretClient 
{
    public override Response<KeyVaultSecret> GetSecret(
        string name,
        string version = null, 
        CancellationToken cancellationToken = default)
        => throw new NotImplementedException();
        
    public override Task<Response<KeyVaultSecret>> GetSecretAsync(
        string name, 
        string version = null, 
        CancellationToken cancellationToken = default)
        => throw new NotImplementedException();
}

SecretClient secretClient = new MockSecretClient();
```

---

### Service client input and output models

Model types hold the data being sent and received from Azure services. There are two main kinds of models:

> * **Input models** are intended to be created and passed as parameters to service methods by developers. They have one or more public constructors and writeable properties.

Reviewer 7: Pure input/output models are the majority case, but we may want to mention "round-trip" models here as well.  These are returned by the service, modified, and used as input.

* **Output models** are only returned by the service and have neither public constructors nor writeable properties.

To create a test instance of an input model use one of the available public constructors and set the additional properties you need.

```csharp
SecretProperties secretProperties = new("secret")
{
    NotBefore = DateTimeOffset.Now
};
```

To create instances of output models, a model factory is used. Most Azure SDK client libraries provide a static model factory class that ends in `ModelFactory` and contains a set of static methods to initialize the library's output model types.

> ```C#

Reviewer 2: ```suggestion
```csharp
```

KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
```

> [!NOTE]

> > Some input models have read-only properties that are only populated when the model is returned by the service. In this case, a model factory method will be available that allows setting these properties. See <xref:Azure.Security.KeyVault.Secrets.SecretModelFactory>.

Reviewer 2: ```suggestion
> Some input models have read-only properties that are only populated when the model is returned by the service. In this case, a model factory method will be available that allows setting these properties. For example, <xref:Azure.Security.KeyVault.Secrets.SecretModelFactory.SecretProperties%2A>.
```

```csharp
// CreatedOn is a read-only property and can only be 

> // set via a model factory SecretProperties.

Reviewer 2: ```suggestion
// set via the model factory's SecretProperties method.
```

secretPropertiesWithCreatedOn = SecretModelFactory.SecretProperties(
    name: "secret", createdOn: DateTimeOffset.Now);
```

## Explore response types

> The <xref:Azure.Response> class is an abstract class that represents an HTTP response and is a part of almost all types returned by service client methods. You can create test `Response` instances using either the Moq library or standard C# inheritence.

Reviewer 4: ```suggestion
The <xref:Azure.Response> class is an abstract class that represents an HTTP response and is returned by most service client methods. You can create test `Response` instances using either the Moq library or standard C# inheritence.
```

## [Moq](#tab/moq)

The Moq library provides concise functionality for setting up mock responses:

```csharp

> Mock responseMock = new Mock();

Reviewer 2: ```suggestion
Mock responseMock = new();
```

Reviewer 6: This is newer syntax I don't think many devs - especially those beginning their journey into C# - will know. I think we should keep the samples simple. This is why in our SDK samples we use neither `var` or `new(...)`. Or, at least, we're not supposed to; admittedly, it's not well-policed.

Reviewer 2: @IEvangelist @alexwolfmsft Should we consider creating an _.editorconfig_ file for Azure SDK code samples stored in the dotnet/docs repo? I'd like to get to place where all code conventions/styles were consistent.

Reviewer: The .NET docs repo uses the https://github.com/dotnet/docs/blob/main/.editorconfig, which is based on the .NET Runtime's config. The idea being that we'd always have our docs reflect theirs and thus be consistent.

> Mock responseMock = new Mock();

Reviewer 7: ```suggestion
Mock<Response> responseMock = new Mock<Response>();
```
The `new` won't compile without the generic included.  If you use `Mock` as the type (sans generic) then you'll need to do an explicit cast on L124

```csharp
Response response = (Response)responseMock.Object;
```

responseMock.SetupGet(r => r.Status).Returns(200);

Response response = responseMock.Object;
```

> To create an instance of Response without defining any behaviors:

Reviewer 2: ```suggestion
To create an instance of `Response` without defining any behaviors:
```

```csharp
Response response = Mock.Of<Response>();
```

## [C#](#tab/csharp)

> The Response class is abstract, which means there are a lot of members to override. Consider using the Moq library to streamline your approach.

Reviewer 2: ```suggestion
The `Response` class is abstract, which means there are a lot of members to override. Consider using the Moq library to streamline your approach.
```

```csharp
public class TestResponse : Response
{
    protected override bool TryGetHeader(string name, out string value)
        => throw new NotImplementedException();
    protected override bool TryGetHeaderValues(string name, out IEnumerable<string> values)
        => throw new NotImplementedException();
         
    protected override bool ContainsHeader(string name) => throw new NotImplementedException();
    protected override IEnumerable<HttpHeader> EnumerateHeaders() 
        => throw new NotImplementedException();

    public override int Status => throw new NotImplementedException();
    public override string ReasonPhrase => throw new NotImplementedException();
    public override Stream ContentStream => throw new NotImplementedException();
    public override string ClientRequestId => throw new NotImplementedException();

    public override void Dispose() => throw new NotImplementedException();
}
```

---

> Some services also support using the <xref:Azure.Response%601> type, which is a class that contains a model and the HTTP response that returned it. To create a test instance of `Response<T>` use the static `Response.FromValue` method:

Reviewer 2: ```suggestion
Some services also support using the <xref:Azure.Response%601> type, which is a class that contains a model and the HTTP response that returned it. To create a test instance of `Response<T>`, use the static `Response.FromValue` method:
```

> Some services also support using the <xref:Azure.Response%601> type, which is a class that contains a model and the HTTP response that returned it. To create a test instance of `Response<T>`, use the static `Response.FromValue` method:

Reviewer 4: @christothes do we need a section on NullableResponse?

Reviewer 3: I would just mention it as a possible return type, but the same principals apply from a mocking perspective.

Reviewer 6: I thought we weren't supposed to return that from any APIs.

Reviewer 3: It should be returned from operations where null is a possible expected success result. Example [here](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/tables/Azure.Data.Tables/api/Azure.Data.Tables.netstandard2.0.cs#L40-L41)

Reviewer: We can always add that later if necessary. Resolving for now.

## [Moq](#tab/moq)

```csharp
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
Response response = Response.FromValue(keyVaultSecret, new TestResponse());
```

## [C#](#tab/csharp)

```csharp
KeyVaultSecret keyVaultSecret = SecretModelFactory.KeyVaultSecret(
    new SecretProperties("secret"), "secretValue");
Response<KeyVaultSecret> response = Response.FromValue(
    keyVaultSecret, Mock.Of<Response>());
```

---

> ### Explore Paging

Reviewer 2: Prefer sentence casing for headings
```suggestion
### Explore paging
```

> The <xref:Azure.Page%601> class is used as a building block in service methods that invoke operations returning results in multiple pages. The `Page<T>` is rarely returned from APIs directly but is useful to create the `AsyncPageable<T>` and `Pageable<T>` instances in the next section. To create a `Page<T>` instance, use the `Page<T>.FromValues` method, passing a list of items, a continuation token, and the Response.

Reviewer 2: ```suggestion
The <xref:Azure.Page%601> class is used as a building block in service methods that invoke operations returning results in multiple pages. The `Page<T>` is rarely returned from APIs directly but is useful to create the `AsyncPageable<T>` and `Pageable<T>` instances in the next section. To create a `Page<T>` instance, use the `Page<T>.FromValues` method, passing a list of items, a continuation token, and the `Response`.
```

AsyncPageable asyncPageable = AsyncPageable.FromPages(new[] { page1, page2, lastPage });
```

## Write a mocked unit test

Suppose your app contains a class that finds the names of keys that will expire within a given amount of time.

```csharp
using Azure.Security.KeyVault.Secrets;

public class AboutToExpireSecretFinder
{
    private readonly TimeSpan _threshold;
    private readonly SecretClient _client;

    public AboutToExpireSecretFinder(TimeSpan threshold, SecretClient client)
    {
        _threshold = threshold;
        _client = client;
    }

    public async Task<string[]> GetAboutToExpireSecrets()
    {
        List<string> secretsAboutToExpire = new();

        await foreach (var secret in _client.GetPropertiesOfSecretsAsync())
        {
            if (secret.ExpiresOn.HasValue &&
                secret.ExpiresOn.Value - DateTimeOffset.Now <= _threshold)
            {
                secretsAboutToExpire.Add(secret.Name);
            }
        }

        return secretsAboutToExpire.ToArray();
    }
}
```

You want to test the following behaviors of the `AboutToExpireSecretFinder` to ensure they continue working as expected:

> * Secrets that don't have an expiry date set are not returned.

Reviewer 2: Rewrite to avoid a double negative. Maybe something like the following:
```suggestion
* Secrets without an expiry date set aren't returned.
```

* Secrets with an expiry date closer to the current date than the threshold are returned.

> When unit testing you only want the unit tests to verify the application logic and not whether the Azure service or SDK works correctly. The following example tests the key behaviors using the popular xUnit framework for C#:

Reviewer 2: Use "library" instead of "SDK" in this case, since we're referring to a single .NET library, not a collection of libraries.
```suggestion
When unit testing you only want the unit tests to verify the application logic and not whether the Azure service or library works correctly. The following example tests the key behaviors using the popular xUnit framework for C#:
```

```csharp
using Azure;
using Azure.Security.KeyVault.Secrets;
using Moq;
using Xunit;

public class AboutToExpireSecretFinderTests
{
    [Fact]
    public async Task DoesNotReturnNonExpiringSecrets()
    {
        // Arrange
        // Create a page of enumeration results
        Page<SecretProperties> page = Page<SecretProperties>.FromValues(new[]
        {
            new SecretProperties("secret1") { ExpiresOn = null },
            new SecretProperties("secret2") { ExpiresOn = null }
        }, null, Mock.Of<Response>());

        // Create a pageable that consists of a single page
        AsyncPageable<SecretProperties> pageable = AsyncPageable<SecretProperties>.FromPages(new[] { page });

        // Setup a client mock object to return the pageable when GetPropertiesOfSecretsAsync is called
        var clientMock = new Mock<SecretClient>();
        clientMock.Setup(c => c.GetPropertiesOfSecretsAsync(It.IsAny<CancellationToken>()))
            .Returns(pageable);

        // Create an instance of a class to test passing in the mock client

>         var finder = new AboutToExpireSecretFinder(TimeSpan.FromDays(2), clientMock.Object);

Reviewer 2: ```suggestion
        AboutToExpireSecretFinder finder = new(TimeSpan.FromDays(2), clientMock.Object);
```

Reviewer 6: 1. I think you shouldn't use `var`. See comment above.
2. I don't think you should use `new(...)`. See same comment above.

This is a fairly basic concept and I think it behooves readers to be as plain as possible - to avoid a lot of newer C#-isms.

Reviewer: This is per our .NET style guides, this conforms with our existing docs - and uses the preferred styling.

        // Act
        var soonToExpire = await finder.GetAboutToExpireSecrets();

        // Assert
        Assert.Empty(soonToExpire);
    }

    [Fact]
    public async Task ReturnsSecretsThatExpireSoon()
    {
        // Arrange

        // Create a page of enumeration results
        DateTimeOffset now = DateTimeOffset.Now;
        Page<SecretProperties> page = Page<SecretProperties>.FromValues(new[]
        {
            new SecretProperties("secret1") { ExpiresOn = now.AddDays(1) },
            new SecretProperties("secret2") { ExpiresOn = now.AddDays(2) },
            new SecretProperties("secret3") { ExpiresOn = now.AddDays(3) }
        }, null, Mock.Of<Response>());

        // Create a pageable that consists of a single page
        AsyncPageable<SecretProperties> pageable = AsyncPageable<SecretProperties>.FromPages(new[] { page });

        // Setup a client mock object to return the pageable when GetPropertiesOfSecretsAsync is called

>         var clientMock = new Mock<SecretClient>();

Reviewer 2: ```suggestion
        Mock<SecretClient> clientMock = new();
```

        clientMock.Setup(c => c.GetPropertiesOfSecretsAsync(It.IsAny<CancellationToken>()))
            .Returns(pageable);

        // Create an instance of a class to test passing in the mock client

>         var finder = new AboutToExpireSecretFinder(TimeSpan.FromDays(2), clientMock.Object);

Reviewer 2: ```suggestion
        AboutToExpireSecretFinder finder = new(TimeSpan.FromDays(2), clientMock.Object);
```

        // Act
        var soonToExpire = await finder.GetAboutToExpireSecrets();

        // Assert
        Assert.Equal(new[] { "secret1", "secret2" }, soonToExpire);
    }
}
```

> ## Refactoring your types for testability

Reviewer: ```suggestion
## Refactor your types for testability
```

    {
        _threshold = threshold;
        _client = new SecretClient(
            new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")),
            new DefaultAzureCredential());
    }
}
```

The simplest refactoring you can do to enable testing with dependency injection would be to expose the client as a parameter and run default creation code when no value is provided. This approach allows you to make the class testable while still retaining the flexibility of using the type without much ceremony.

```csharp
public class AboutToExpireSecretFinder
{
    public AboutToExpireSecretFinder(TimeSpan threshold, SecretClient client = null) 
    { 
        _threshold = threshold; 
        _client = client ?? new SecretClient(
            new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")),
            new DefaultAzureCredential());
    }
}
```

Another option is to move the dependency creation entirely into the calling code:

```csharp
public class AboutToExpireSecretFinder
{
    public AboutToExpireSecretFinder(TimeSpan threshold, SecretClient client)
    {
        _threshold = threshold;
        _client = client;
    }
}

> var secretClient = new SecretClient(

Reviewer 2: ```suggestion
SecretClient secretClient = new(
```

    new Uri(Environment.GetEnvironmentVariable("KeyVaultUri")), 
    new DefaultAzureCredential());
var finder = new AboutToExpireSecretFinder(TimeSpan.FromDays(2), secretClient);
```

This approach is useful when you would like to consolidate the dependency creation and share the client between multiple consuming classes.

## See also

* [Dependency injection in .NET](/dotnet/core/extensions/dependency-injection)
* [Unit testing best practices](/dotnet/core/testing/unit-testing-best-practices)

> * [Unit testing C# in .NET Core using dotnet test and xUnit](/dotnet/core/testing/unit-testing-with-dotnet-test)

Reviewer 2: The linked article title should be updated to remove "Core" too.
```suggestion
* [Unit testing C# in .NET using dotnet test and xUnit](/dotnet/core/testing/unit-testing-with-dotnet-test)
```
